### DISCUSSION TITLE
PRG 패턴에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#200] [Network]
뭘까용
Post Redirect Get

**Comment by alstn113** [comment]
PRG 패턴(PRG Pattern)은 웹 애플리케이션 개발에서 POST 요청 후에 리다이렉트를 통해 중복 폼 제출을 방지하는 패턴입니다. PRG은 Post-Redirect-Get의 약어로, 간단한 웹 폼 작업에서 유용하게 사용됩니다. 이 패턴은 사용자 경험을 향상시키고, 데이터의 중복 제출을 방지하며, 브라우저의 "뒤로 가기" 버튼 사용 시에도 예상대로 동작하도록 합니다.

### 동작 원리

1. **폼 제출 (POST 요청)**:
    - 사용자가 폼을 제출하면 일반적으로 POST 요청이 서버로 전송됩니다. 이 때 서버는 폼 데이터를 받아 처리하고, 그 결과를 응답으로 클라이언트에게 전송합니다.
2. **리다이렉트 (Redirect)**:
    - 서버는 POST 요청을 처리한 후, 클라이언트에게 새로운 경로(예: 성공 또는 실패 페이지)로 리다이렉트하도록 응답합니다. 이 과정에서 서버는 클라이언트가 다시 요청을 보내는 것을 유도합니다.
3. **중복 제출 방지 (Get 요청)**:
    - 클라이언트는 서버에서 받은 리다이렉트 응답에 따라 GET 요청을 보냅니다. 이 GET 요청은 폼 제출과는 별개로, 사용자가 이전 페이지로 돌아갈 수 있도록 합니다.

### 예시

1. 사용자가 로그인 폼에 유효하지 않은 정보를 입력하고 제출합니다.
2. 서버는 POST 요청을 처리하고, 유효하지 않은 입력을 감지하면 로그인 실패 페이지로 리다이렉트합니다.
3. 사용자는 로그인 실패 페이지를 볼 수 있지만, 브라우저의 "뒤로 가기" 버튼을 누르면 다시 로그인 폼으로 돌아갈 수 있습니다.

**Comment by 3Juhwan** [comment]
PRG 패턴은 Post/Redirect/Get 패턴의 약자로, 웹 애플리케이션에서 폼 제출 후 페이지 새로 고침이나 브라우저 뒤로 가기 등의 문제를 방지하기 위해 사용하는 디자인 패턴입니다. 일반적으로 멱등성이 보장되지 않는 POST 요청에 사용합니다. 

PRG 패턴은 다음과 같은 단계를 따릅니다. 

1. Post: 사용자가 폼을 제출하면 클라이언트는 서버에 POST 요청을 보냅니다. 서버는 이 요청을 처리하여 데이터베이스를 업데이트하거나 다른 작업을 수행합니다.
2. Redirect: 서버는 POST 요청을 처리한 후, 클라이언트에게 새로운 URL로 리디렉션하라는 응답을 보냅니다. 이 리디렉션은 클라이언트에게 302 Found 상태 코드와 함께 새로운 URL을 포함한 Location 헤더를 반환하여 수행됩니다.
3. Get: 클라이언트는 서버의 응답을 받아 새로운 URL로 GET 요청을 보냅니다. 서버는 이 GET 요청을 처리하여 최종 결과 페이지를 클라이언트에게 반환합니다. 

PRG 패턴을 따르지 않으면 발생하는 문제는 다음과 같습니다. 

1. 폼을 제출(POST)해도 redirection을 하지 않아 새로운 페이지로 이동하지 않습니다. 
2. 이 상태에서 페이지를 새로고침하면 이전과 동일한 POST 요청이 다시 발생하여 멱등성이 보장되지 않을 수 있습니다.
